### ISSUE TITLE
ENH: LinAlg Symmetric Matrix Class

### ISSUE THREAD

**Issue by carlos-adir** [numpy/numpy#20021]
### Proposed new feature or change:

In this [Stackoverflow's link](https://stackoverflow.com/questions/2572916/numpy-smart-symmetric-matrix) they discuss implementing ```symmetric 2D matrix class``` inherit from ```numpy.ndarray```, which takes half of the space it would normally take.
It would be nice to have this class cause some methods (exemple ```numpy.linalg.inv```) would be faster and wouldn't make unnecessary verifications, along with getting less numerical error on floating point's operations.
Another advantage (that affects the speed) is that we can use theorems for these matrices and algorithms like [LDL Cholesky Decomposition](https://en.wikipedia.org/w/index.php?title=Cholesky_decomposition&oldid=1046948641) to find ```eigenvalues``` faster.

**Comment by seberg** [comment]
Thanks for the request/forward!  My first guess is, that may have more luck (and a better fit) with sparse array implementations (such as `pydata.sparse`).  Since it seems a bit similar to a sparse format to me (albeit, maybe one quite different from typical sparse matrices).

>  inherit from `numpy.ndarray`, which takes half of the space it would normally take.

Unfortunately, such a class cannot inherit from `ndarray`, since it is not compatible with the `ndarray` memory model.  Because of that, I expect it would be far easier to kick-start such an effort outside outside of NumPy.  

Good integration with NumPy should be achievable using the `__array_ufunc__` and `__array_function__` protocols.  If you were to start such a project within NumPy, that would be the best/first way to approach it as well!

If you *would* aim for implementing as much as possible in C (i.e. not using a single NumPy array/memory chunk for storage):  I would love to close the gap, so that you can use NumPy C functionality without having to back your data as an `ndarray`!  For example, recent restructures in the ufuncs and casting could be refined a bit further and made public to allow custom types to access a good part of NumPy's core functionality without using NumPy arrays for storage.
I actually do think that would be a very interesting (and timely) area to work on within NumPy, it might allow interesting applications.  But I am not even sure it would help for this request (or has any real-world use-cases).  For your diagonal array, backing with a normal numpy array seems like a good (enough) fit.  (This type of thing could be a game-changer for objects that would require _many_ small arrays for data storage)

Maybe someone will disagree, but right now I am not sure we can provide anything besides a discussion forum and some feedback.  I certainly don't think that NumPy currently has significant resources to throw at such a project (there would be a few other projects I would rank higher).
But, maybe you can find some wider interest, and you may get lucky and maybe someone involved here can help you shake the right tree, connect with someone interested enough to write grant, etc.  (although this issue tracker is probably not the best place either).

As such, however, I think this is not something that we can put on our roadmap/agenda.  At least not at this time.

**Comment by carlos-adir** [comment]
Thank you for the answer! The application is mainly for space matrix for Finite Element Method.
I'm still not sure to not use ```numpy.ndarray```, cause my initial idea is to storage the data in 1D-array and modify the access M[i, j] to get the right position, like
```python
def __getitem__(self, (i, j)):
    if i < j:
        i, j = j, i
    return self.__M[j + i*(i+1)/2]
```
I don't know exactly how ```numpy.ndarray``` is implemented in low-level, but with that we avoid coding again low-level functions like ```__add__```, ```__sub__``` arrays and we profit that the sum of two SymmetricMatrix is the sum of these 1D-array.

I will make some code and tests in python to verify if I can use ```np.ndarray``` before using C and I will let you know of my progress. 
Last question: Should I be concerned about the "format" of coding by now? I would like to join the code into numpy project someday so other people could use it. But I'm not sure if I will have to code all again to fits numpy's standard format.
It's my first time contributing to a big project such as numpy

**Comment by mhvk** [comment]
Comment from the sidelines: I think often the factor 2 memory gain is not such a big deal, and I think it is relatively hard to get right (e.g., one obviously should also be able to get out rows and columns, for which the simple scheme you give above would fail; and for complex matrices, one generally cares about Hermitian ones).

But the use of suitable linear algebra routines can give a real speed-up. So, one could start by still writing a new class that automatically uses the right linear algebra functions (using `__array_function__`).

**Comment by seberg** [comment]
Indeed, I was probably a bit too specific to the request rather than to the problem :).  It is also unclear to me the best solution to the problem (using fast linear algebra) is actually helped much by implementing such a class (or at least a full featured one).

E.g. there were also discussion in SciPy to add a function for fast detection whether a dense array is (almost) symmetrical and to use symmetrical linear algebra functions in that case.  There may be arguments for including such a function in NumPy or SciPy.  That may be the far more relevant thing to pursue here!  Or as Marten suggest: write a class that automatically does this, but still uses a a dense array as backing.

If you really want to start such a project, you can aim for inclusion in NumPy to define a direction to work in.  But it is unlikely to ever happen, such a project should aim to be part of the ecosystem around numpy/scipy, but there is probably no reason for it to be _in_ NumPy.  There _might_ be projects where it fits, which is why I mentioned sparse.  But I am not sure.
